src/data/polynomial/degree/reverse_mul.lean

```diff
@@ -0,0 +1,209 @@
+def rev_at (N : ℕ) : ℕ → ℕ := λ i : ℕ , ite (i ≤ N) (N-i) i
+
+@[simp] lemma rev_at_invol {N n : ℕ} : rev_at N (rev_at N n) = n :=
+begin
+  unfold rev_at,
+  split_ifs,
+    { exact nat.sub_sub_self h, },
+    { exfalso,
+      apply h_1,
+      exact nat.sub_le N n, },
+    { refl, },
+end
+
+@[simp] lemma rev_at_small {N n : ℕ} (H : n ≤ N) : rev_at N n = N-n :=
+begin
+  unfold rev_at,
+  split_ifs,
+  refl,
+end
+
+/-- reflect of a natural number N and a polynomial f, applies the function rev_at to the exponents of the terms appearing in the expansion of f.  In practice, reflect is only used when N is at least as large as the degree of f.  Eventually, it will be used with N exactly equal to the degree of f.  -/
+def reflect : ℕ → polynomial R → polynomial R := λ N : ℕ , λ f : polynomial R , ⟨ (rev_at N '' ↑(f.support)).to_finset , λ i : ℕ , f.coeff (rev_at N i) , begin
+  simp_rw [set.mem_to_finset, set.mem_image, mem_coe, mem_support_iff],
+  intro,
+  split,
+    { intro a_1,
+      rcases a_1 with ⟨ a , ha , rfl⟩,
+      rwa rev_at_invol, },
+    { intro,
+      use (rev_at N a),
+      rwa [rev_at_invol, eq_self_iff_true, and_true], },
+end ⟩
+
+@[simp] lemma reflect_zero {n : ℕ} : reflect n (0 : polynomial R) = 0 :=
+begin
+  refl,
+end
+
+@[simp] lemma reflect_add {f g : polynomial R} {n : ℕ} : reflect n (f+g) = reflect n f + reflect n g :=
+begin
+  ext1,
+  unfold reflect,
+  simp only [coeff_mk, coeff_add],
+end
+
+@[simp] lemma reflect_smul (N : ℕ) {r : R} : reflect N (C r * f) = C r * (reflect N f) :=
+begin
+  ext1,
+  unfold reflect,
+  simp only [coeff_mk, coeff_C_mul],
+end
+
+@[simp] lemma reflect_C_mul_X_pow (N n : ℕ) {c : R} : reflect N (C c * X ^ n) = C c * X ^ (rev_at N n) :=
+begin
+  ext1,
+  unfold reflect,
+  rw coeff_mk,
+  by_cases h : rev_at N n = n_1,
+    { rw [h, coeff_C_mul, coeff_C_mul, coeff_X_pow_self, ← h, rev_at_invol, coeff_X_pow_self], },
+    { rw not_mem_support_iff_coeff_zero.mp,
+        { symmetry,
+          apply not_mem_support_iff_coeff_zero.mp,
+          intro,
+          apply h,
+          exact (mem_support_C_mul_X_pow a).symm, },
+        { intro,
+          apply h,
+          rw ← @rev_at_invol N n_1,
+          apply congr_arg _,
+          exact (mem_support_C_mul_X_pow a).symm, }, },
+end
+
+@[simp] lemma reflect_monomial (N n : ℕ) : reflect N ((X : polynomial R) ^ n) = X ^ (rev_at N n) :=
+begin
+  rw [← one_mul (X^n), ← one_mul (X^(rev_at N n)), ← C_1, reflect_C_mul_X_pow],
+end
+
+/-- The reverse of a polynomial f is the polynomial obtained by "reading f backwards".  Even though this is not the actual definition, reverse f = f (1/X) * X ^ f.nat_degree. -/
+def reverse : polynomial R → polynomial R := λ f , reflect f.nat_degree f
+
+lemma nat_degree_add_of_mul_leading_coeff_nonzero (f g: polynomial R) (fg: f.leading_coeff * g.leading_coeff ≠ 0) :
```

Comment:
This is pretty much `nat_degree_mul'`

Reply:
Thank you so much!  Indeed, ```nat_degree_mul' fg``` works as a term proof of the lemma!  I will remove the lemma and references to it and replace them by nat_degree_mul'.